R/modspec.R

```diff
@@ -39,6 +39,68 @@ write_capture <- function(x) {
   paste0("_capture_[",i-1,"] = ", x[i], ";") 
 }
 
+#' This function adds ETA values to the capture list when the `@etas` 
+#' option is used in `$CAPTURE` .
+#' 
+#' @param x the model object
+#' @param env the mread parse environment
+#' 
+#' @details
+#' We look at env$capture_etas to see if there were any expression text passed
+#' through the `@etas` option. This text will be parsed and evaluated in the 
+#' model environment after adding `LAST` and `last` to represent the last
+#' ETA (or the [total] number of rows in `$OMEGA`).
+#' 
+#' An error is generated in case the expression can't be parsed and evaluated.
+#' 
+#' `@etas` must resolve to an integer-like object. Expecting most usage to be
+#' `1:last` which will be integer, but we want to support `c(1,2,5)` as well
+#' which will not be integer. 
+#'
+#' @return The model object, possibly updated.   
+#' 
+#' @noRd
+capture_etas <- function(x, env) {
+  if(!is.character(env[["capture_etas"]])) return(x)
+  last <- dim_matlist(omat(x))
+  if(last==0) return(x)
+  parse_env <- as.list(env$ENV)
+  parse_env$last <- parse_env$LAST <- last
+  for(eta_txt in env[["capture_etas"]]) {
+    etan <- try(eval(parse(text = eta_txt), envir = parse_env), silent = TRUE)
+    if(inherits(etan, "try-error")) {
+      msg <- c(
+        glue("could not parse this expression for `etas`: {eta_txt}."), 
+        x = etan
+      )
+      abort(msg)
+    }
+    resolves_int <- 
+      is.integer(etan) || 
+      (is.numeric(etan) && all(abs(etan - round(etan)) < 1e-08))
+    if(!resolves_int) {
+      abort("`etas` must resolve to an integer value.")    
+    }
+    if(length(etan)==0) {
+      abort("`etas` has length 0.")  
+    }
+    etan <- unique(as.integer(round(etan)))
+    if(any(etan > last)) {
+      abort(
+        message = c(
+          glue("`etas` must be integers between 1 and {last}."),
+          i = glue("minimum value in `etas`: {min(etan)}"),
+          i = glue("maximum value in `etas`: {max(etan)}")
```

Comment:
Should this check the lower bound too (i.e. that no values are below 1)?

```diff
diff --git a/inst/maintenance/unit/test-capture.R b/inst/maintenance/unit/test-capture.R
index 90647e5f..7d04ea49 100644
--- a/inst/maintenance/unit/test-capture.R
+++ b/inst/maintenance/unit/test-capture.R
@@ -139,4 +139,10 @@ test_that("capture with @etas directive", {
     mcode("capture-at-etas-7", code), 
     regexp = "has length 0"
   )
+
+  code <- paste0(base, "-1:5")
+  expect_error(
+    mcode("capture-at-etas-8", code),
+    regexp = "must be integers between 1 and 5"
+  )
 })
```

Reply:
Thanks; added the unit test and updated the check in the code.